File: tmf2/org.lttng.scope.tmf2.views.core/src/org/lttng/scope/tmf2/views/core/timegraph/model/provider/states/TimeGraphModelStateProvider.java
/*
 * Copyright (C) 2017 EfficiOS Inc., Alexandre Montplaisir <alexmonthy@efficios.com>
 *
 * All rights reserved. This program and the accompanying materials are
 * made available under the terms of the Eclipse Public License v1.0 which
 * accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 */

package org.lttng.scope.tmf2.views.core.timegraph.model.provider.states;

import java.util.List;

import org.eclipse.jdt.annotation.Nullable;
import org.eclipse.tracecompass.tmf.core.trace.ITmfTrace;
import org.lttng.scope.tmf2.views.core.timegraph.model.render.StateDefinition;

import com.google.common.collect.ImmutableList;

import javafx.beans.property.ObjectProperty;
import javafx.beans.property.SimpleObjectProperty;

/**
 * Basic implementation of {@link ITimeGraphModelStateProvider}.
 *
 * @author Alexandre Montplaisir
 */
public abstract class TimeGraphModelStateProvider implements ITimeGraphModelStateProvider {

    private final ObjectProperty<@Nullable ITmfTrace> fTraceProperty = new SimpleObjectProperty<>(null);

    private final List<StateDefinition> fStateDefinitions;

    /**
     * Constructor
     *
     * @param stateDefinitions
     *            The state definitions used in this provider
     */
    public TimeGraphModelStateProvider(List<StateDefinition> stateDefinitions) {
        fStateDefinitions = ImmutableList.copyOf(stateDefinitions);
    }

    @Override
    public ObjectProperty<@Nullable ITmfTrace> traceProperty() {
        return fTraceProperty;
    }

    @Override
    public List<StateDefinition> getStateDefinitions() {
        return fStateDefinitions;
    }

}

File: jabberwocky-core/src/main/kotlin/com/efficios/jabberwocky/views/timegraph/model/provider/states/TimeGraphModelStateProvider.kt
/*
 * Copyright (C) 2017 EfficiOS Inc., Alexandre Montplaisir <alexmonthy@efficios.com>
 *
 * All rights reserved. This program and the accompanying materials are
 * made available under the terms of the Eclipse Public License v1.0 which
 * accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 */

package com.efficios.jabberwocky.views.timegraph.model.provider.states;

import com.efficios.jabberwocky.common.TimeRange;
import com.efficios.jabberwocky.project.TraceProject;
import com.efficios.jabberwocky.views.timegraph.model.provider.ITimeGraphModelProvider;
import com.efficios.jabberwocky.views.timegraph.model.render.StateDefinition;
import com.efficios.jabberwocky.views.timegraph.model.render.states.TimeGraphStateRender;
import com.efficios.jabberwocky.views.timegraph.model.render.tree.TimeGraphTreeElement;
import com.efficios.jabberwocky.views.timegraph.model.render.tree.TimeGraphTreeRender;
import javafx.beans.property.ObjectProperty;
import javafx.beans.property.SimpleObjectProperty

import java.util.concurrent.FutureTask;
import java.util.function.Function;
import java.util.stream.Collectors;

/**
 * Provider for timegraph state intervals.
 *
 * It can be used stand-alone (ie, for testing) but usually would be part of a
 * {@link ITimeGraphModelProvider}.
 *
 * @author Alexandre Montplaisir
 */
abstract class TimeGraphModelStateProvider(val stateDefinitions: List<StateDefinition>) {

    /**
     * Property representing the trace this arrow provider fetches its data for.
     */
    private val traceProjectProperty: ObjectProperty<TraceProject<*, *>?> = SimpleObjectProperty(null)
    fun traceProjectProperty() = traceProjectProperty
    var traceProject
        get() = traceProjectProperty.get()
        set(value) = traceProjectProperty.set(value)

    /**
     * Get a state render for a single tree element
     *
     * @param treeElement
     *            The tree element to target
     * @param timeRange
     *            The time range of the query
     * @param resolution
     *            The resolution, in timestamp units, of the query. For example,
     *            a resolution of 1000 means the provider does not need to
     *            return states that don't cross at least 1 1000-units borders.
     * @param task
     *            An optional task parameter, which can be checked for
     *            cancellation to stop processing at any point and return.
     * @return The corresponding state render for this tree element and settings
     */
    abstract fun getStateRender(treeElement: TimeGraphTreeElement,
                                timeRange: TimeRange,
                                resolution: Long,
                                task: FutureTask<*>?): TimeGraphStateRender

    open fun getStateRenders(treeElements: Set<TimeGraphTreeElement>,
                        timeRange: TimeRange,
                        resolution: Long,
                        task: FutureTask<*>?): Map<TimeGraphTreeElement, TimeGraphStateRender> {
        return treeElements.associate { Pair(it, getStateRender(it, timeRange, resolution, task)) }
    }

    /**
     * Helper method to fetch all the state renders for all tree elements of a
     * *tree* render.
     *
     * Default implementation simply calls {@link #getStateRender} on each tree
     * element sequentially, but more advanced providers could override it if
     * they can provide a faster mechanism.
     *
     * @param treeRender
     *            The tree render for which to prepare all the state renders
     * @param timeRange
     *            The time range of the query
     * @param resolution
     *            The resolution of the query, see {@link #getStateRender}.
     * @param task
     *            The optional task parameter which can be used for
     *            cancellation.
     * @return The corresponding state renders
     */
    open fun getAllStateRenders(treeRender: TimeGraphTreeRender,
                           timeRange: TimeRange,
                           resolution: Long,
                           task: FutureTask<*>?): List<TimeGraphStateRender> {
        return treeRender.allTreeElements.map { getStateRender(it, timeRange, resolution, task) }
    }

}
